network/src/application/interface.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::application::{
+    db::PeerDb,
+    types::{PeerError, PeerInfo, PeerState},
+};
+use async_trait::async_trait;
+use diem_types::PeerId;
+use std::collections::{hash_map::Entry, HashMap};
+
+/// A generic `NetworkInterface` for applications to connect to networking
+#[async_trait]
+pub trait NetworkInterface {
+    type Sender;
+
+    /// Provides the PeerDb for other functions.  Not expected to be used externally.
```

Comment:
It's to avoid code duplication, since this will be exactly the same code.  I'm basically trying to work around inheritance, which maybe I shouldn't be doing in rust.